tests/ui/unusable_matches_binding.stderr

```diff
@@ -0,0 +1,42 @@
+error: using identificator (`unusable_binding`) as matches! pattern argument without guard matches all cases
```

Comment:
Let's be brief for error messages. I think something like

```
error: binding `unusable_binding` is not used
```

Then we can explain more in detail in `note` or `help`

Reply:
The current message is a bit verbose, I agree, but in my opinion that suggested message isn't really much better because rustc already emits a warning here with an almost identical message (unused variable: `unusable_binding`), so I don't think that the suggested message would provide much more value than what rustc already says about this.

The problem imo is not so much that the binding isn't used, but rather that the user didn't mean to create a binding in the first place, and someone that doesn't know how `matches!` works likely won't understand why a binding is created either.
The user more likely meant to compare two values (I've seen this confusion happen a few times in forums, thinking that `matches!(1, x)` behaves like `1 == x`), so I think it'd be better if this message focused on the fact that it won't do that and that this `matches!` is useless/always evaluates to true.

How about something like
```
error: identifier pattern in `matches!` macro always evaluates to true

note: the identifier pattern matches any value and creates an unusable binding in the process
help: if you meant to compare two values, use `x == y` or `discriminant(x) == discriminant(y)`
```

Reply:
> and someone that doesn't know how matches! works likely won't understand why a binding is created either.

Fair point

> rustc already emits a warning here with an almost identical message (unused variable: unusable_binding)

Regarding this, I wonder if we should have this lint in Clippy at all. Users would stumble into rustc's lint first before running clippy. Having this as rustc_lint might be more beneficial

Reply:
I've changed the error message here.
Isn't this lint too "specific" to be included in the `rustc`?

Reply:
I agree that it's a bit "unfortunate" that rustc's unused_variables lint triggers here and prevents us from emitting better diagnostics. It could do a much better job at explaining what the issue is and how to fix it. Might make sense to specialize the behavior of the `unused_variables` lint, if it sees that the parent node of the identifier pattern is a `matches!` expansion?

OTOH, the lint is in the correctness category (deny by default), so I think that will stand out more than rustc's lint? I'm not sure.

Reply:
I guess it's better to have something than nothing. Let's see how this lint fares, maybe if it's good enough we can have it uplifted to rustc lint